【Feature】棒グラフ機能の実装

masa5281 opened the pull request:
## 概要
本PRでは消費カロリーの推移を確認できる棒グラフ機能を実装しました。
グラフは月間と週間の2種類を採用しており、タブで切り替えが可能です。
使用ライブラリは「Recharts」です。
## 主な変更点
- **棒グラフコンポーネントの実装**
  - 週グラフ
  - 月グラフ
- **タブの切り替え処理**
- **Tailwind CSSによるスタイリング**
- **モジュール化・コンポーネント化**
## 関連するIssues
Close #18 
※詳細なタスクについてはIssueをご確認ください。
## レビュワーへのお願い
下記項目のご確認をお願いいたします。
- 月間・週間のグラフが正しく表示されるか
- タブによる表示切り替えが正しく動作するか
- 冗長または不要なコードがないか

masa5281 committed:
feat: 棒グラフページのルーティングを定義（#18）

masa5281 committed:
feat: Rechartsを導入（#18）

masa5281 committed:
feat: Caloriesコントローラのindexアクションとルーティングを定義（#18）

masa5281 committed:
feat: カロリー取得APIとカスタムフックを定義（#18）

masa5281 committed:
feat: 棒グラフで今週、前週分のカロリーを表示（#18）

masa5281 committed:
feat: 週グラフで翌週分のカロリーを表示（#18）

masa5281 committed:
feat: 【週グラフ】グラフ選択後のスタイルを変更（#18）

masa5281 committed:
fix: 【週グラフ】翌週ボタン押下時に月曜日のカロリーを取得するよう修正（#18）

masa5281 committed:
feat: グラフのラベルを実装（#18）

masa5281 committed:
fix: 【週グラフ】先週ボタン押下後のグラフ選択でカロリーが反映されない問題を解決（#18）

masa5281 committed:
feat: 【月グラフ】月グラフを実装（#18）

masa5281 committed:
refactor: 共通のグラフコンポーネントを抽出（#18）

masa5281 committed:
refactor: グラフの構造部分をコンポーネント化（#18）

masa5281 committed:
refactor: グラフの期間ボタンを共通化（#18）

masa5281 committed:
refactor: 日付周りの処理をモジュール化（#18）

masa5281 committed:
chore: コメントの修正と空行を整理（#18）

yama-mountain commented on frontend/src/components/BarGraph/CalorieBarGraph.jsx:
```diff
@@ -0,0 +1,71 @@
+  dataKey,
+}) => {
+  return (
+    <ResponsiveContainer className={""} width="100%" height={350} >
```
空のclassName指定不要そう？

masa5281 replied:
すみません。
削除をし忘れていました！
修正いたしました。

[修正後のコミット](https://github.com/masa5281/nanko-meshi/pull/19/commits/24478b665e60cbee70c79d751ac8cb2715e6e3ee)

yama-mountain commented on frontend/src/components/BarGraph/MonthGraph.jsx:
```diff
@@ -0,0 +1,90 @@
+    setSelectCalorie(data.burnedCalorie);
+  };
+
+  useEffect(() => {
+    const endYearDate = new Date(startYearDate.getFullYear(), 11, 31);
+    setIsCurrentYear(formatToday >= startYearDate && formatToday <= endYearDate);
+  }, [startYearDate]);
+
+  // DBのカロリーデータからグラフを生成
+  useEffect(() => {
+    const data = [];
+
+    for (let i = 0; i < 12; i++) {
+      const baseDate = new Date(startYearDate);
+      baseDate.setMonth(baseDate.getMonth() + i);
+      const dateStr = `${baseDate.getFullYear()}-${String(baseDate.getMonth() + 1).padStart(2, "0")}`;
+      const newCalorieList = calorieList.filter(calorie => calorie.recorded_at.slice(0, 7) === dateStr);
+      const sumCalorie = newCalorieList.reduce((totalCalorie, object) => totalCalorie + object.burned_calorie, 0);
+      if (i === selectMonth.getMonth()) {
+        setSelectCalorie(sumCalorie);
+      }
+      data.push({
+        month: month[i],
+        burnedCalorie: sumCalorie
+      })
+    }
+    setBarData(data);
+  }, [startYearDate, calorieList, selectMonth, month]);
```
この辺、useEffectを利用する理由は何ですか？

masa5281 replied:
下記のタイミングでDBから月ごとの集計をやり直し、その結果を`setBarDate`で再レンダリングさせるためです。
DBから取得した日付の文字列と直接比較しているため、複雑になっています。

1. 初回描画時
2. 前年、翌年選択時（`startYearDate`の変更時）
3. グラフクリック時（`selectMonth`の変更時）

yama-mountain replied:
useEffectは副作用を扱うHookなので、そのような用途で使うべきではないです！
（このレポジトリの他の既存コードもそうですが……）

一つの案ですが、useStateを使うことをオススメします。
この一連の処理を `buildBarData(calorieList, selectMonth, startYearDate)` 関数に外だしすることができたと仮定した時

barData stateの初期化でまず1をクリアします

```javascript
const [barData, setBarData] = useState(buildBarData(calorieList, selectMonth, startYearDate);
```

2と3はそれぞれの変更時に(具体的には `setStartYearDate` や `setSelectMonth` を動かしているところ）でsetBarDatします

```javascript
  const onPrevYear = () => {
    const prevYearDate = new Date(startYearDate.getFullYear() - 1, 0, 1);
    setStartYearDate(prevYearDate);
    setSelectMonth(prevYearDate);
    setBarData(buildBarData(calorieList, prevYearDate, prevYearDate)
  };
```

useEffectを「なにかの値が変わったら自動的に同期してくれるもの」と理解するのはNGです！

masa5281 replied:
グラフ作成ロジックと日付計算処理、不要なuseStateを整理しリファクタリングしてみました。
コミットが1つになってしまい申し訳ないです。

for文とpushに関しては今回はそのまま採用させていただきましたが、
次回以降での参考にさせていただきます。
よろしくお願いいたします。

[修正後のコミット](https://github.com/masa5281/nanko-meshi/commit/9e0f314749dd3def191ae18b96fcb93301514392)

yama-mountain commented on frontend/src/components/BarGraph/WeekGraph.jsx:
```diff
@@ -0,0 +1,96 @@
+  const [barData, setBarData] = useState([]);
+  const [isCurrentWeek, setIsCurrentWeek] = useState(true);
+  const [startDate, setStartDate] = useState(new Date(currentYear, currentMonth, mondayDate));
+  const week = useMemo(() => ["月", "火", "水", "木", "金", "土", "日"], []);
```
更新されることがないならuseMemo不要そう

masa5281 replied:
useMemoを削除しました！

yama-mountain commented on frontend/src/components/BarGraph/MonthGraph.jsx:
```diff
@@ -0,0 +1,90 @@
+    setIsCurrentYear(formatToday >= startYearDate && formatToday <= endYearDate);
+  }, [startYearDate]);
+
+  // DBのカロリーデータからグラフを生成
```
懸念されていた共通化ですが、確かに必要そうです。全体的に読みづらく何がしたいのかがぱっと分かりづらいです（そもそも下部で聞いている通りでuseEffectを使うのも良くないので！）

まず検討して欲しいのが

- dayjsなどの日付系ライブラリを導入して、自前で日付をゴリゴリいじるのをやめる
  - せめて変換系処理をlib的なところにまとめる
- for文をやめる
  - Arrayのdataにpushして情報を組み上げるのをやめる。他でreduceが使えているので同様にできるはず
- 特定の月の情報か？をnewCalorieListとして積み上げているが、そこの比較を文字列で行うのをやめる
  - ここもdayjsとかであれば簡単にいけそう
- という処理を「描画コンポーネント」であるMonthGraphやWeeklyGraphに入れない
  - 純粋関数の形で表せるように、他の所に関数を定義する
  - 今回であれば、「baseDate, calorieList, weekly/monthly」を渡したら「 `{month, burnedCalorie}[]` 」を返す、みたいな関数を用意したい

masa5281 committed:
refactor: 日付操作とグラフデータ生成ロジックを切り出し（#18）

- utils/graphData.js: buildBarData関数で週・月グラフのデータを構築
- utils/graphDate.js: フォーマット・範囲判定・日付、年数加算関数を分離

masa5281 committed:
refactor: 不要なclassNameを削除（#18）
